### PR TITLE
Resolve #1

### LINKED ISSUE #1: Review 1
* https://github.com/unitprotocol/wrapped-assets/blob/2723ad8f3e5d6baad35f5177c12d5f94f6c541ae/contracts/WrappedAsset.sol#L22-L30
Will work incorrectly if actual transferred value differs from requested (e.g. fees, etc).

* https://github.com/unitprotocol/wrapped-assets/blob/2723ad8f3e5d6baad35f5177c12d5f94f6c541ae/contracts/WrappedAsset.sol#L41
Not sure if a space in `symbol` is a good idea. Some third-party soft will definitely puke )

Moreover, is it worth to push these into ERC20 ctor https://github.com/unitprotocol/wrapped-assets/blob/2723ad8f3e5d6baad35f5177c12d5f94f6c541ae/contracts/WrappedAsset.sol#L36-L42 ?

And also store `ERC20(underlying).decimals()` locally?

### PR DESCRIPTION
**Pull request by 34x4p08**
Resolve #1

### CODE REVIEW

File: contracts/WrappedAsset.sol [hunk 1]
```diff
@@ -20,13 +20,17 @@ contract WrappedAsset is ERC20 {
     }
 
     function wrap(uint amount) external {
-        _mint(msg.sender, amount);
+        uint balanceBefore = underlying.balanceOf(address(this));
         underlying.safeTransferFrom(msg.sender, address(this), amount);
+        uint balanceAfter = underlying.balanceOf(address(this));
```
Comment (Eenae) [inline contracts/WrappedAsset.sol:39]: If the `underlying` gives some dividends to `this` before actual balance update, they are attributed to `msg.sender`. That's bad.

File: contracts/WrappedAsset.sol [hunk 2]
```diff
@@ -20,13 +20,17 @@ contract WrappedAsset is ERC20 {
...
+        _mint(msg.sender, balanceAfter.sub(balanceBefore));
     }
 
     function unwrap(uint amount) external {
-        _burn(msg.sender, amount);
+        uint balanceBefore = underlying.balanceOf(address(this));
         underlying.safeTransfer(msg.sender, amount);
+        uint balanceAfter = underlying.balanceOf(address(this));
```
Comment (Eenae) [inline contracts/WrappedAsset.sol:46]: If the `underlying` withhold some fee from `this` on transfer, the fee is not charged from `msg.sender` but stays on his balance, meaning we become a bit insolvent. That's bad.

File: contracts/WrappedAssetFactory.sol [hunk 3]
```diff
@@ -11,8 +11,13 @@ contract WrappedAssetFactory is Ownable {
 
     event NewWrapper(address indexed underlying, address wrapper);
 
-    function deploy(address _underlying) external onlyOwner {
-        WrappedAsset wa = new WrappedAsset(_underlying, wrappers[_underlying].length);
+    function deploy(
+        address _underlying,
+        string calldata _symbolOverride,
+        string memory _nameOverride,
```
Comment (Eenae) [inline contracts/WrappedAssetFactory.sol:17]: `calldata`?

File: contracts/WrappedAsset.sol [hunk 4]
```diff
@@ -58,4 +76,8 @@ contract WrappedAsset is ERC20 {
         }
         str = string(bstr);
     }
+
+    function _isEmptyString(string memory a) public pure returns (bool) {
+        return (keccak256(abi.encodePacked((a))) == keccak256(abi.encodePacked((""))));
```
Comment (Eenae) [inline contracts/WrappedAsset.sol:81]: `bytes(a).length == 0`?

File: contracts/WrappedAsset.sol [hunk 5]
```diff
@@ -12,33 +12,51 @@ contract WrappedAsset is ERC20 {
...
+    ) ERC20(_nameOverride, _symbolOverride) {
         underlying = ERC20(_underlying);
-        _wrapperId = __wrapperId;
+        _symbol =  _isEmptyString(_symbolOverride)
+            ? _concat("W-", _uint2str(_wrapperId), "-", ERC20(_underlying).symbol())
+            : _symbolOverride;
+        _name = _isEmptyString(_nameOverride)
+            ? _concat("Wrapped-", _uint2str(_wrapperId), " ", ERC20(_underlying).name())
+            : _nameOverride;
+        _decimals = _decimalsOverride == 0 ? ERC20(_underlying).decimals() : _decimalsOverride;
```
Comment (Eenae) [inline contracts/WrappedAsset.sol:33]: Here there is no way to override decimals to `0` )

File: contracts/WrappedAsset.sol [hunk 6]
```diff
@@ -12,33 +12,51 @@ contract WrappedAsset is ERC20 {
...
+        uint8 _decimalsOverride
+    ) ERC20(_nameOverride, _symbolOverride) {
         underlying = ERC20(_underlying);
-        _wrapperId = __wrapperId;
+        _symbol =  _isEmptyString(_symbolOverride)
+            ? _concat("W-", _uint2str(_wrapperId), "-", ERC20(_underlying).symbol())
+            : _symbolOverride;
+        _name = _isEmptyString(_nameOverride)
+            ? _concat("Wrapped-", _uint2str(_wrapperId), " ", ERC20(_underlying).name())
+            : _nameOverride;
```
Comment (Eenae) [inline contracts/WrappedAsset.sol:32]: As an alternative c++ style solution, the base ctor may be invoked with results from `internal view` functions which properly construct attributes. So there is no need in shadowing slots.

**Review by Eenae — CHANGES_REQUESTED** [review]
Transfers handling should be improved.

File: contracts/WrappedAsset.sol (hunk 1 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol:39]: Can you provide an example of this use case?

File: contracts/WrappedAsset.sol (hunk 6 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol]: Moved to factory

File: contracts/WrappedAssetFactory.sol (hunk 3 above)
Comment (34x4p08) [reply contracts/WrappedAssetFactory.sol]: +

File: contracts/WrappedAsset.sol (hunk 4 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol]: +

File: contracts/WrappedAsset.sol (hunk 5 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol]: Agree

File: contracts/WrappedAsset.sol (hunk 2 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol]: Fixed

File: contracts/WrappedAsset.sol (hunk 1 above)
Comment (Eenae) [reply contracts/WrappedAsset.sol:30]: https://github.com/mixbytes/solidity/blob/master/contracts/token/DividendToken.sol#L69

File: contracts/WrappedAsset.sol (hunk 2 above)
Comment (Eenae) [reply contracts/WrappedAsset.sol]: What if the actual transfer amount is greater that `amount`?

File: contracts/WrappedAsset.sol (hunk 1 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol:30]: - We shouldn't worry about super rare use cases to not overload wrapper code
- It's unwanted to use dividend assets as collateral on Unit, compounding assets are preferred instead

File: contracts/WrappedAsset.sol (hunk 2 above)
Comment (34x4p08) [reply contracts/WrappedAsset.sol]: Sounds like incorrect transfer logic, so we will not use such assets as collateral

File: contracts/WrappedAsset.sol (hunk 1 above)
Comment (Eenae) [reply contracts/WrappedAsset.sol:30]: Btw, vulnerable to reentrancy if `underlying.transferFrom` allows external calls.

Comment (34x4p08) [reply contracts/WrappedAsset.sol:31]: ReentrancyGuard added

**Review by Eenae — APPROVED** [review]
Ok, with the non-standard tokens proviso.

**Pull request merged**